### ISSUE TITLE
Adding same node twice creates perpetual loop

### ISSUE THREAD

**Issue by lukehinds** [keylime/python-keylime#75]
# Environment
 * OS / version: Fedora 29
 * Processor architecture: x86_64
 * TPM Manufacturer:  emulator
 * Keylime version: master

# Description 

Add a node twice creates a perpetual loop

## Expected behavior vs. actual behavior 

Keylime_verifier and keylime_tenant should report this only once, and then break / exit the loop.

## Steps to reproduce problem

Run the following command:

`keylime_tenant -v 127.0.0.1 -t 127.0.0.1 -f somefile --uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000`

Run the same again:

```
2019-03-08 13:34:45.752 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.761 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
2019-03-08 13:34:45.771 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.780 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
2019-03-08 13:34:45.789 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.799 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
2019-03-08 13:34:45.808 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.818 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
2019-03-08 13:34:45.827 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.837 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
2019-03-08 13:34:45.846 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.856 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
2019-03-08 13:34:45.865 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.875 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
2019-03-08 13:34:45.884 - keylime.cloudverifier - INFO - DELETE returning 200 response for instance id: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2019-03-08 13:34:45.894 - keylime.cloudverifier - WARNING - Node of uuid D432FBB3-D2F1-4A97-9EF7-75BD81C00000 already exists
```

```
2019-03-08 13:34:44.995 - keylime.tenant - WARNING - Node already existed at CV.  Deleting and re-adding...
2019-03-08 13:34:45.004 - keylime.tenant - INFO - Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 deleted from CV
2019-03-08 13:34:45.004 - keylime.tenant - DEBUG - b64_v:bsoXgk2ceWMVn42XlXiLhOPtzCRAMQ1fbVwCwCyqQcs=
2019-03-08 13:34:45.013 - keylime.tenant - WARNING - Node already existed at CV.  Deleting and re-adding...
2019-03-08 13:34:45.023 - keylime.tenant - INFO - Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 deleted from CV
2019-03-08 13:34:45.023 - keylime.tenant - DEBUG - b64_v:bsoXgk2ceWMVn42XlXiLhOPtzCRAMQ1fbVwCwCyqQcs=
2019-03-08 13:34:45.032 - keylime.tenant - WARNING - Node already existed at CV.  Deleting and re-adding...
2019-03-08 13:34:45.042 - keylime.tenant - INFO - Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 deleted from CV
2019-03-08 13:34:45.042 - keylime.tenant - DEBUG - b64_v:bsoXgk2ceWMVn42XlXiLhOPtzCRAMQ1fbVwCwCyqQcs=
2019-03-08 13:34:45.052 - keylime.tenant - WARNING - Node already existed at CV.  Deleting and re-adding...
2019-03-08 13:34:45.061 - keylime.tenant - INFO - Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 deleted from CV
2019-03-08 13:34:45.061 - keylime.tenant - DEBUG - b64_v:bsoXgk2ceWMVn42XlXiLhOPtzCRAMQ1fbVwCwCyqQcs=
2019-03-08 13:34:45.070 - keylime.tenant - WARNING - Node already existed at CV.  Deleting and re-adding...
2019-03-08 13:34:45.080 - keylime.tenant - INFO - Node D432FBB3-D2F1-4A97-9EF7-75BD81C00000 deleted from CV
2019-03-08 13:34:45.080 - keylime.tenant - DEBUG - b64_v:bsoXgk2ceWMVn42XlXiLhOPtzCRAMQ1fbVwCwCyqQcs=
```

**Comment by nabilschear** [comment]
yes, i've seen this before.  it's a nasty problem that has to do with threading, python, and synchronization.  Tornado has multiple processes with their own event loops (see https://github.com/keylime/python-keylime/blob/master/keylime/cloud_verifier_tornado.py#L507 ), so data synchronization happens through sqlite.  when a node already exists the tenant code tries to delete and re-add it (https://github.com/keylime/python-keylime/blob/master/keylime/tenant.py#L481 ).  Because the synchronization happens lazily in the verifier, it takes it a moment to register the node deleted and the tenant keeps trying to add it again.

I see a few potential fixes for this:

1. properly fix the race condition (ewww, may take a large amount of re-factoring)
2. set `multiprocessing_pool_num_workers` in `keylime.conf` to 1.  can't have a race when there's only one
3. sleep a little after the call to delete to let it persist before trying to re-add
4. raise an error and tell the user they must delete first before adding (this really just masks the race)

I'm partial to number 4.  if you call status on the node as it's being deleted, it will tell you that it's been terminated and is starting to go away.  it just takes it once through the cloud verifier state machine to actually get removed from the database.

anyone have thoughts on this?  @jetwhiz @lukehinds

**Comment by nabilschear** [comment]
another option i just thought of: the tenant code can keep some state that it's trying to delete it and suppress the error messages until it is done.  Probably need a small sleep to keep the tenant from polling too quickly.  So, this is in essence a pretty version of number 3 above.  but instead of a random sleep, it's a polling sleep.

**Comment by lukehinds** [comment]
I initially though of 3 / 4, but the problem then is what happens if there are more then one tenant (or more than one entity making the call to add). This is why fixing this in the verifier might be a better approach. I don't know the Rest API to well, but I guess the same call is available there and if we patch up the tenant, then the same issue would still reside over an API call to add? 

We could perhaps use a sort of token bucket in the verifier?

**Comment by lukehinds** [comment]
@nabilschear @jetwhiz , so to summarise, unless there is a specific need or something I have not considered, can we fix this by removing:

https://github.com/keylime/python-keylime/blob/7b0a5b4c57e0efd7add2cac88a7ad0eca95e9a90/keylime/tenant.py#L482-L483

And let the user call with the args `delete` and `add` and make the decision themselves?

**Comment by nabilschear** [comment]
the intent behind the delete/re-add pattern was to allow a user to update an existing node.  if some of the metadata changes (i.e., new pcr values), this was the mechanism to handle this.  

what about this plan?

I agree that add on an existing node shouldn't update it.  i'll update the code to raise an error if you try to add to an existing node.

to more explicitly handle this, I'll add an update command that will do the delete/re-add pattern.  though we also have re-activate which actually does change the node in place.  

and finally i think that delete should poll the verifier to confirm the node was deleted before returning.  this will keep the race from happening even if someone else calls keylime directly and does delete re-add

**Comment by lukehinds** [comment]
sounds good to me @nabilschear , thank you!

**Comment by lukehinds** [comment]
I grabbed all of the logs as my environment is really unstable / unpredictable and I am not sure if I still undstand how things should behave.

The commands used were:

    keylime_tenant -v 127.0.0.1 -t 127.0.0.1 -u D432FBB3-D2F1-4A97-9EF7-75BD81C00000 -c delete
    keylime_tenant -v 127.0.0.1 -t 127.0.0.1 -u D432FBB3-D2F1-4A97-9EF7-75BD81C00000 -f somefile -c add

Logs:

https://gist.github.com/lukehinds/c832697109b552c1d7776d96560b2766
https://gist.github.com/lukehinds/4341434641db5ec0b053e8f430b6db51
https://gist.github.com/lukehinds/4d4c22104e1984109934e598930790a8
